client.go

```diff
@@ -0,0 +1,289 @@
+func runOnce[TREQ any, TRESP commonResponseSetter](
+	ctx context.Context,
+	logger *logs.Logger,
+	treq *request[TREQ],
+) (
+	tresp *response[TRESP],
+	err error,
+) {
+	if err = treq.client.cfg.ratelim.Wait(ctx); err != nil {
+		return
+	}
+
+	// request body
+	var reqBody []byte
+	if treq.body != nil {
+		reqBody, err = json.Marshal(treq.body)
+		if err != nil {
+			err = retry.PermanentError{Cause: err}
+			return
+		}
+	}
+
+	// request
+	req, err := http.NewRequestWithContext(ctx, treq.method, requestURL(treq),
+		bytes.NewReader(reqBody))
+	if err != nil {
+		err = retry.PermanentError{Cause: err}
+		return
+	}
+
+	// headers
+	mergeHeaders(req.Header, treq.headers)
+
+	// credentials
+	err = treq.client.creds.configure(
+		logger.SubLogger(logs.WithPrefix("Authorization")),
+		req)
+	if err != nil {
+		return // allow retry
+	}
+
+	// send the request
+	resp, err := treq.client.cfg.httpClient.Do(req)
+	if err != nil {
+		// errors from Do() may be permanent or not, it is not possible to
+		// determine precisely
+		return // allow retry
+	}
+
+	// handle response
+	if resp.StatusCode >= 400 {
+		err = handleErrorResponse(resp, logger)
+		logFullRequestError(logger, treq, reqBody, err)
+		return
+	}
+
+	tresp, err = handleSuccessResponse[TRESP](resp, logger)
+	if err != nil {
+		logFullRequestError(logger, treq, reqBody, err)
+		return
+	}
+
+	if treq.client.cfg.logSuccess {
+		logFullHTTPRequestSuccess(logger, treq, reqBody, tresp)
+	}
+
+	return tresp, err
+}
+
+func handleSuccessResponse[TRESP commonResponseSetter](httpResp *http.Response, logger *logs.Logger) (
+	*response[TRESP],
+	error,
+) {
+	var ret response[TRESP]
+
+	ret.code = httpResp.StatusCode
+	ret.headers = httpResp.Header
+
+	var err error
+	ret.rawBody, err = io.ReadAll(httpResp.Body)
+	if err != nil {
+		// allow retry
+		return nil, errors.Join(err, HTTPError{
+			Code:    httpResp.StatusCode,
+			RawBody: ret.rawBody,
+			Headers: ret.headers,
+		})
+	}
+
+	err = json.Unmarshal(ret.rawBody, &ret.body)
+	if err != nil {
+		// allow retry
+		return nil, errors.Join(err, HTTPError{
+			Code:    httpResp.StatusCode,
+			RawBody: ret.rawBody,
+			Headers: ret.headers,
+		})
+	}
+
+	return &ret, nil
+}
+
+func handleErrorResponse(resp *http.Response, logger *logs.Logger) error {
+	// the error response must always support errors.As(err, HTTPError)
+	httpErr := HTTPError{
+		Code:    resp.StatusCode,
+		Headers: resp.Header,
+	}
+
+	respBody, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("CloudFlare returned an error, but failed to read the error body: %v; %w", err, httpErr) // allow retry
+	}
+
+	httpErr.RawBody = respBody
+
+	// try to parse the CloudFlare error objects
+	mediaType, _, err := mime.ParseMediaType(resp.Header.Get("content-type"))
+	if err != nil {
+		return retry.PermanentError{Cause: fmt.Errorf("CloudFlare returned an error, and the content-type of the response is unsupported %q (%v); %w",
```

Comment:
```suggestion
		return retry.PermanentError{Cause: fmt.Errorf("CloudFlare returned an error, and the content-type of the response is invalid %q (%v); %w",
```

unsupported sounds like cfdns does not support it

Reply:
We only support json :). But I changed it.